<a href="https://colab.research.google.com/github/leekate/2020-ctrc-OralCancer-Classification-project/blob/master/Oral_Detect_my_ResNet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

!pip uninstall keras
!pip install Keras==2.2.4

!pip uninstall tensorflow
!pip install tensorflow==1.13.1

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 원본 데이터 파일 갯수 확인

In [12]:
import os

ori_length1= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/train/1.Cancer"
ori_length2= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/validation/1.Cancer"
ori_plus1= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/test/1.Cancer"

ori_file_list1 = os.listdir(ori_length1)
ori_file_list2 = os.listdir(ori_length2)
ori_plus_list1 = os.listdir(ori_plus1)

print("cancer")
print(len(ori_file_list1))
print(len(ori_file_list2))
print(len(ori_plus_list1))

cancer
124
143
143


In [13]:
ori_length3= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/train/2.Precancer"
ori_length4= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/validation/2.Precancer"
ori_plus2= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/test/2.Precancer"

ori_file_list3 = os.listdir(ori_length3)
ori_file_list4 = os.listdir(ori_length4)
ori_plus_list2 = os.listdir(ori_plus2)

print("precancer")
print(len(ori_file_list3))
print(len(ori_file_list4))
print(len(ori_plus_list2))

precancer
46
52
52


In [14]:
ori_length5 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/train/3.Inflammatory"
ori_length6 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/validation/3.Inflammatory"
ori_plus3= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/test/3.Inflammatory"

ori_file_list5 = os.listdir(ori_length5)
ori_file_list6 = os.listdir(ori_length6)
ori_plus_list3 = os.listdir(ori_plus3)

print("extra")
print(len(ori_file_list5))
print(len(ori_file_list6))
print(len(ori_plus_list3))

extra
79
93
93


In [15]:
ori_length7 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/train/4.Normal"
ori_length8 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/validation/4.Normal"
ori_plus4= "/content/drive/Shareddrives/CTRC-OralDetect-Project/Original/test/4.Normal"

ori_file_list7 = os.listdir(ori_length7)
ori_file_list8 = os.listdir(ori_length8)
ori_plus_list4 = os.listdir(ori_plus4)

print("normal")
print(len(ori_file_list7))
print(len(ori_file_list8))
print(len(ori_plus_list4))

normal
455
341
341


In [16]:
import pandas as pd

#df = df_maker(3,4,0)  # 3X3 데이터 프레임 생성 0으로 채움
col = ["Train image","Validation Image","Test Image","Total Image"]
ind = ["Cancer","Precancer","Extra","Normal","TOTAL"]
#df.columns = col  # 컬럼 col 리스트로 덮어씌움
#df.index = ind    # 인덱스 ind 리스트로 덮어씌움


con = [
       [len(ori_file_list1),len(ori_file_list2),len(ori_plus_list1),len(ori_plus_list1)+len(ori_file_list1)+len(ori_file_list2)], 

       [len(ori_file_list3),len(ori_file_list4),len(ori_plus_list2),len(ori_plus_list2)+len(ori_file_list3)+len(ori_file_list4)],

       [len(ori_file_list5),len(ori_file_list6),len(ori_plus_list3),len(ori_plus_list3)+len(ori_file_list5)+len(ori_file_list6)],

       [len(ori_file_list7),len(ori_file_list8),len(ori_plus_list4),len(ori_plus_list4)+len(ori_file_list7)+len(ori_file_list8)],

        [len(ori_file_list1)+len(ori_file_list3)+len(ori_file_list5)+len(ori_file_list7),
         len(ori_file_list2)+len(ori_file_list4)+len(ori_file_list6)+len(ori_file_list8),
         len(ori_plus_list1)+len(ori_plus_list2)+len(ori_plus_list3)+len(ori_plus_list4),
        len(ori_file_list1)+len(ori_file_list3)+len(ori_file_list5)+len(ori_file_list7)+len(ori_file_list2)+len(ori_file_list4)+len(ori_file_list6)+len(ori_file_list8)+len(ori_plus_list1)+len(ori_plus_list2)+len(ori_plus_list3)+len(ori_plus_list4)]
       ]

pd.DataFrame(con,columns=col, index=ind)

Train image  Validation Image  Test Image  Total Image
Cancer             124               143         143          410
Precancer           46                52          52          150
Extra               79                93          93          265
Normal             455               341         341         1137
TOTAL              704               629         629         1962

# 단순 증량 후 최종 파일 갯수 확인

In [17]:
length1= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/train/1.Cancer"
length2= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/validation/1.Cancer"
plus1= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/test/1.Cancer"

file_list1 = os.listdir(length1)
file_list2 = os.listdir(length2)
plus_list1 = os.listdir(plus1)

print("cancer")
print(len(file_list1))
print(len(file_list2))
print(len(plus_list1))

cancer
496
143
143


In [18]:
length3= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/train/2.Precancer"
length4= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/validation/2.Precancer"
plus2= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/test/2.Precancer"

file_list3 = os.listdir(length3)
file_list4 = os.listdir(length4)
plus_list2 = os.listdir(plus2)

print("precancer")
print(len(file_list3))
print(len(file_list4))
print(len(plus_list2))

precancer
184
52
52


In [19]:
length5 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/train/3.Inflammatory"
length6 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/validation/3.Inflammatory"
plus3= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/test/3.Inflammatory"

file_list5 = os.listdir(length5)
file_list6 = os.listdir(length6)
plus_list3 = os.listdir(plus3)

print("extra")
print(len(file_list5))
print(len(file_list6))
print(len(plus_list3))

extra
316
93
93


In [20]:
length7 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/train/4.Normal"
length8 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/validation/4.Normal"
plus4= "/content/drive/Shareddrives/CTRC-OralDetect-Project/OriginDataGenerate/test/4.Normal"

file_list7 = os.listdir(length7)
file_list8 = os.listdir(length8)
plus_list4 = os.listdir(plus4)

print("normal")
print(len(file_list7))
print(len(file_list8))
print(len(plus_list4))

normal
1365
341
341


In [28]:
import pandas as pd

#df = df_maker(3,4,0)  # 3X3 데이터 프레임 생성 0으로 채움
col = ["Train image","Validation Image","Test Image","Total Image"]
ind = ["Cancer","Precancer","Extra","Normal","TOTAL"]
#df.columns = col  # 컬럼 col 리스트로 덮어씌움
#df.index = ind    # 인덱스 ind 리스트로 덮어씌움


con = [
       [len(file_list1),len(file_list2),len(plus_list1),len(plus_list1)+len(file_list1)+len(file_list2)], 
       [len(file_list3),len(file_list4),len(plus_list2),len(plus_list2)+len(file_list3)+len(file_list4)],
       [len(file_list5),len(file_list6),len(plus_list3),len(plus_list3)+len(file_list5)+len(file_list6)],
       [len(file_list7),len(file_list8),len(plus_list4),len(plus_list4)+len(file_list7)+len(file_list8)],

        [len(file_list1)+len(file_list3)+len(file_list5)+len(file_list7),
         len(file_list2)+len(file_list4)+len(file_list6)+len(file_list8),
         len(plus_list1)+len(plus_list2)+len(plus_list3)+len(plus_list4),
        len(file_list1)+len(file_list3)+len(file_list5)+len(file_list7)+len(file_list2)+len(file_list4)+len(file_list6)+len(file_list8)+len(plus_list1)+len(plus_list2)+len(plus_list3)+len(plus_list4)]
       ]

pd.DataFrame(con,columns=col, index=ind)

Train image  Validation Image  Test Image  Total Image
Cancer             496               143         143          782
Precancer          184                52          52          288
Extra              316                93          93          502
Normal            1365               341         341         2047
TOTAL             2361               629         629         3619

# Histogram Equalization 한 데이터 양

In [22]:
length1= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/train/1.Cancer"
length2= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/validation/1.Cancer"
plus1= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/test/1.Cancer"

file_list1 = os.listdir(length1)
file_list2 = os.listdir(length2)
plus_list1 = os.listdir(plus1)

print("cancer")
print(len(file_list1))
print(len(file_list2))
print(len(plus_list1))

cancer
496
143
143


In [23]:
length3= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/train/2.Precancer"
length4= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/validation/2.Precancer"
plus2= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/test/2.Precancer"

file_list3 = os.listdir(length3)
file_list4 = os.listdir(length4)
plus_list2 = os.listdir(plus2)

print("precancer")
print(len(file_list3))
print(len(file_list4))
print(len(plus_list2))

precancer
184
52
52


In [24]:
length5 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/train/3.Inflammatory"
length6 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/validation/3.Inflammatory"
plus3= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/test/3.Inflammatory"

file_list5 = os.listdir(length5)
file_list6 = os.listdir(length6)
plus_list3 = os.listdir(plus3)

print("extra")
print(len(file_list5))
print(len(file_list6))
print(len(plus_list3))

extra
316
93
93


In [25]:
length7 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/train/4.Normal"
length8 = "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/validation/4.Normal"
plus4= "/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/test/4.Normal"

file_list7 = os.listdir(length7)
file_list8 = os.listdir(length8)
plus_list4 = os.listdir(plus4)

print("normal")
print(len(file_list7))
print(len(file_list8))
print(len(plus_list4))

normal
1365
341
341


In [27]:
import pandas as pd

#df = df_maker(3,4,0)  # 3X3 데이터 프레임 생성 0으로 채움
col = ["Train image","Validation Image","Test Image","Total Image"]
ind = ["Cancer","Precancer","Extra","Normal","TOTAL"]
#df.columns = col  # 컬럼 col 리스트로 덮어씌움
#df.index = ind    # 인덱스 ind 리스트로 덮어씌움


con = [
       [len(file_list1),len(file_list2),len(plus_list1),len(plus_list1)+len(file_list1)+len(file_list2)], 
       [len(file_list3),len(file_list4),len(plus_list2),len(plus_list2)+len(file_list3)+len(file_list4)],
       [len(file_list5),len(file_list6),len(plus_list3),len(plus_list3)+len(file_list5)+len(file_list6)],
       [len(file_list7),len(file_list8),len(plus_list4),len(plus_list4)+len(file_list7)+len(file_list8)],

        [len(file_list1)+len(file_list3)+len(file_list5)+len(file_list7),
         len(file_list2)+len(file_list4)+len(file_list6)+len(file_list8),
         len(plus_list1)+len(plus_list2)+len(plus_list3)+len(plus_list4),
        len(file_list1)+len(file_list3)+len(file_list5)+len(file_list7)+len(file_list2)+len(file_list4)+len(file_list6)+len(file_list8)+len(plus_list1)+len(plus_list2)+len(plus_list3)+len(plus_list4)]
       ]

pd.DataFrame(con,columns=col, index=ind)

Train image  Validation Image  Test Image  Total Image
Cancer             496               143         143          782
Precancer          184                52          52          288
Extra              316                93          93          502
Normal            1365               341         341         2047
TOTAL             2361               629         629         3619

# 본격적인 설계 시작
## ResNet

In [3]:
from tensorflow.keras.applications import ResNet50

conv_base=ResNet50(weights='imagenet',
                include_top=False,
                input_shape=(150,150,3))

94773248/94765736 [==============================] - 1s 0us/step


In [4]:
from tensorflow.keras import layers
from tensorflow.keras import models
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout
from tensorflow.keras import optimizers, initializers, regularizers, metrics



model=models.Sequential()
model.add(conv_base)



model.add(layers.Dense(2048, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(2048, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add( MaxPooling2D((2,2), padding='same'))




model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(1024, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))




model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(512, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))





model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(256, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))









model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))
model.add(layers.Dense(128, kernel_regularizer = regularizers.l2
                                  (0.001),activation='relu'))




model.add(Flatten()) 
model.add(layers.Dense(64,activation='relu'))
model.add(layers.Dense(4,activation='relu'))


model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [5]:
conv_base.trainable=False
#weight가 업데이트되지 않도록

In [6]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
resnet50 (Functional)        (None, 5, 5, 2048)        23587712  
_________________________________________________________________
dense (Dense)                (None, 5, 5, 2048)        4196352   
_________________________________________________________________
dense_1 (Dense)              (None, 5, 5, 2048)        4196352   
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 3, 3, 2048)        0         
_________________________________________________________________
dense_2 (Dense)              (None, 3, 3, 1024)        2098176   
_________________________________________________________________
dense_3 (Dense)              (None, 3, 3, 1024)        1049600   
_________________________________________________________________
dense_4 (Dense)              (None, 3, 3, 1024)        1

## Training

In [14]:
from keras.models import load_model
import os
# model checkpoint
model_sav_path = '/content/drive/Shareddrives/CTRC-OralDetect-Project/jiwoo_model/ResNet50'

if not os.path.isdir(model_sav_path):
    os.mkdir(model_sav_path)
    
model_path = model_sav_path + 'my_Resnet_model.h5'


In [9]:
from tensorflow.keras.callbacks import EarlyStopping
import tensorflow.keras
early_stopping = EarlyStopping()

callbacks=[tensorflow.keras.callbacks.TensorBoard(
    log_dir='my_log_dir',
    histogram_freq=1,
    embeddings_freq=1,),
    early_stopping]


In [15]:
train_dir='/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/train'
validation_dir='/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/validation'
test_dir='/content/drive/Shareddrives/CTRC-OralDetect-Project/NewDataGenerate-HM/test'

In [16]:
import os
from tensorflow.keras.preprocessing.image import ImageDataGenerator,array_to_img,img_to_array, load_img
import numpy as np

np.random.seed(3)

imageGenerator = ImageDataGenerator(rescale=1./255)
trainGen = imageGenerator.flow_from_directory(train_dir,
                                              target_size=(150,150))
                                            

validationGen = imageGenerator.flow_from_directory(validation_dir,
                                              target_size=(150,150))
                                              
testGen = imageGenerator.flow_from_directory(test_dir,
                                              target_size=(150,150))
                                            
                                    

Found 2361 images belonging to 4 classes.
Found 629 images belonging to 4 classes.
Found 629 images belonging to 4 classes.


* steps_per_epoch : 한 epoch에 사용한 스텝 수를 지정합니다. 총 2361개의 훈련 샘플이 있고 배치사이즈가 32이므로 73 스텝으로 지정합니다.
* validation_data : 검증데이터셋을 제공할 제네레이터를 지정합니다. 여기서는 앞서 생성한 validation_generator으로 지정합니다.
* validation_steps : 한 epoch 종료 시 마다 검증할 때 사용되는 검증 스텝 수를 지정합니다. 총 629개의 검증 샘플이 있고 배치사이즈가 32이므로 19 스텝으로 지정합니다.
* epochs : 전체 훈련 데이터셋에 대해 학습 반복 횟수를 지정합니다. 50번을 반복적으로 학습시켜 보겠습니다.


In [17]:
history=model.fit_generator(trainGen,steps_per_epoch=73, epochs=50,
                  validation_data=validationGen,validation_steps=19,
                           callbacks=callbacks)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/training.py:1844: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  warnings.warn('`Model.fit_generator` is deprecated and '


Epoch 1/50
73/73 [==============================] - 769s 11s/step - loss: 15.6179 - accuracy: 0.5766 - val_loss: 12.0684 - val_accuracy: 0.5477
Epoch 2/50
73/73 [==============================] - 32s 436ms/step - loss: 10.3771 - accuracy: 0.6376 - val_loss: 9.1785 - val_accuracy: 0.5395
Epoch 3/50
73/73 [==============================] - 32s 439ms/step - loss: 8.0698 - accuracy: 0.6505 - val_loss: 7.4602 - val_accuracy: 0.5378
Epoch 4/50
73/73 [==============================] - 32s 432ms/step - loss: 6.4683 - accuracy: 0.6861 - val_loss: 6.4322 - val_accuracy: 0.5395
Epoch 5/50
73/73 [==============================] - 32s 434ms/step - loss: 5.4929 - accuracy: 0.6690 - val_loss: 5.2678 - val_accuracy: 0.5395
Epoch 6/50
73/73 [==============================] - 32s 434ms/step - loss: 4.8358 - accuracy: 0.5771 - val_loss: 4.5286 - val_accuracy: 0.5378
Epoch 7/50
73/73 [==============================] - 32s 429ms/step - loss: 4.1945 - accuracy: 0.5801 - val_loss: 3.9762 - val_accuracy: 0.53

In [1]:
from keras.utils import plot_model
from keras.models import load_model
model = load_model('/content/drive/Shareddrives/CTRC/jiwoo_model/ResNet50/my_Resnet_model.h5')
plot_model(model,show_shapes=True,to_file='my_Resnet_model.png') 

# Testing

In [ ]:
model= load_model('/content/drive/Shareddrives/CTRC/jiwoo_model/Resnet50/my_Resnet_model.h5')

In [ ]:
import matplotlib.pyplot as plt

#predict=model.predict(#validationData의 feature)
predict=model.predict(testGen)

## Result

### 정확도 및 손실 그래프

In [ ]:
# import matplotlib.pyplot as plt


# acc = history.history['accuracy']
# val_acc = history.history['val_accuracy']
# loss = history.history['loss']
# val_loss = history.history['val_loss']

# epochs = range(len(acc))

# plt.plot(epochs, acc, 'r', label='Training acc')
# plt.plot(epochs, val_acc, 'b', label='Validation acc')
# plt.title('Training and validation accuracy')
# plt.legend()

# plt.figure()

# plt.plot(epochs, loss, 'r', label='Training loss')
# plt.plot(epochs, val_loss, 'b', label='Validation loss')
# plt.title('Training and validation loss')
# plt.legend()

# plt.show()




import matplotlib.pyplot as plt


acc = history.history['acc']
val_acc = history.history['val_acc']
loss = history.history['loss']
val_loss = history.history['val_loss']

epochs = range(len(acc))

plt.plot(epochs, acc, 'r', label='Training acc')
plt.plot(epochs, val_acc, 'b', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()

plt.figure()

plt.plot(epochs, loss, 'r', label='Training loss')
plt.plot(epochs, val_loss, 'b', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()

plt.show()

# 알고리즘 CNN
* 심층 신경망(DNN, Deep Neural Network)
* 합성곱 신경망(CNN, Convoultional Neural Network)
* 순환 신경망(RNN, Recurrent Neural Network)
* 제한 볼츠만 머신(RBM, Restricted Boltzmann Machine)
* 심층 신뢰 신경망(DBN, Deep Belief Network)



CNN. 합성곱신경망 [Convolutional Neural Network]

CNN은 하나 이상의 합성곱 계층과 그 위에 올려진 일반적인 인공 신경망 계층들로 이루어져 있고, 가중치와 레이어들을 추가로 활용한다. 이러한 구조로 CNN은 2차원 구조의 입력 데이터를 충분히 활용할 수 있다. 다른 딥러닝 구조들과 비교했을 때, CNN은 영상과 음성 분야 모두에서 좋은 성능을 보인다. 또한 표준 역전달을 통해 훈련될 수 있고, 적은 수의 매개변수를 사용한다는 이점이 있다. 

 CNN은 마지막 레이어에 손실 함수를 가지며, 우리가 일반 신경망을 학습시킬 때 사용하던 각종 기법들을 동일하게 적용할 수 있다. CNN은 이미지 및 비디오 인식, 추천 시스템 및 자연 언어 처리 등에 응용된다.

# 텐서플로우의 시각화 프레임워크

In [ ]:
from keras.utils import plot_model
plot_model(model, to_file='model.png')

In [ ]:
%load_ext tensorboard
%tensorboard --logdir=my_log_dir

# 최대 epoch
```
early_stopping = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=50)
```
성능이 증가하지 않는다고, 그 순간 바로 멈추는 것은 효과적이지않을 수 있다. patience 는 성능이 증가하지 않는 epoch 을 몇 번이나 허용할 것인가를 정의


# ROC & 정밀도 및 재현율

In [ ]:
from sklearn.metrics import roc_curve

cm = confusion_matrix(validationGen, predict)
print(cm)
print( "accuracy : ", accuracy_score( validationGen, predict ) ) # (cm[0][0]+cm[1][1])/(cm[0][0]+cm[1][1]+cm[1][0]+cm[0][1])
print( "recall(detection rate) : ", recall_score( y_test, predict ) )
print( "fallout (false alarm rate) : ",cm[1][0]/(cm[1][0]+cm[1][1]) )
print( "f1_score : ", f1_score(validationGen, predict) )
print( "roc_auc : ", roc_auc_score(validationGen, predict) )

fpr, tpr, threshold = roc_curve(validationGen, predict)

plt.plot(fpr, tpr, 'o-', label="ROC Curve")
plt.plot([0,1],[0,1], 'k--')
fallout = cm[1][0]/(cm[1][0]+cm[1][1])
plt.plot([fallout], [recall_score( validationGen, predict )], 'ro', ms=10)
plt.xlabel("fallout")
plt.ylabel("recall")
plt.title("ROC curve")
plt.show()
